### PR TITLE
Release `v0.3.9`

### PR DESCRIPTION
**Pull request by Urhengulas**
This PR prepares the release of `defmt v0.3.9` and related crates.

There are two breaking changes, one in `defmt-decoder`, one in `defmt-parser`. This does not affect users of the firmware `defmt` crate, but only of the host-side `defmt-decoder` crate. The breakage is minimal though. We are adding one new variant to two enums which were not `non_exhaustive` before and those enums are not returned from a function but are used for configuration.

This release also no longer requires the "unstable" feature from the decoder and parser (the flag remains, but is a no-op). They have been stable for years now (except one accidental breakage).

| Crate             | Old Version | New Version | Breaking Change | Target |
| ----------------- | ----------- | ----------- | --------------- |--------|
| defmt             | 0.3.8       | 0.3.9       | -               | 🎙️  |
| defmt-macros      | 0.3.9       | 0.3.10      | -               | 🎙️  |
| defmt-decoder     | 0.3.11      | 0.4.0       | 💥              | 🖥️ |
| defmt-parser      | 0.3.4       | 0.4.0       | 💥              | 🎙️/🖥️ |
| defmt-print       | 0.3.12      | 0.3.13      | -               | 🖥️ |
| defmt-semihosting | 0.1.0       | 0.1.0       | N/A              | 🎙️ |
| qemu-run          | 0.1.0       | 0.0.0       | N/A              | 💻 |
| xtask             | 0.1.0       | 0.0.0       | N/A              | 💻 |

### CODE REVIEW

**Comment by Urhengulas** [comment]
Opened https://github.com/probe-rs/probe-rs/pull/2809. Once that is reviewed and considered good I will make the release.

File: CHANGELOG.md
Comment (jonathanpallant) [inline CHANGELOG.md:8]: This should be the same heading level as the versions under it.

Comment (Urhengulas) [reply CHANGELOG.md:8]: defmt releases are h2, all others are h3. Fixed it.

**Comment by jonathanpallant** [comment]
On my list to look at this week

File: README.md
Comment (jonathanpallant) [inline README.md:25]:
```suggestion
The minimum supported Rust version is 1.76, which translates to Ferrocene 24.05.
```

**Comment by jonathanpallant** [comment]
This PR contains a bunch of substantial changes, as well as the release version number bumps. I'd like to see the substantial changes brought out into a separate PR so they can be tested and reviewed individually.

For example https://github.com/knurling-rs/defmt/pull/864/commits/3ba66f7338d7d641af099065faadecc50a860d0b requires a change in CI to both build with `--locked` and without.

https://github.com/knurling-rs/defmt/pull/864/commits/b5ed24c5b99baf88333eccc26788d56b2ee71c3b also requires a change in CI to actually do the build on the MSRV (be that Rust 1.76 or Ferrocene).

**Comment by jonathanpallant** [comment]
I tested the radio-app exercises using this new defmt:

```diff
diff --git a/nrf52-code/radio-app/Cargo.toml b/nrf52-code/radio-app/Cargo.toml
index 44d2206..52d3e65 100644
--- a/nrf52-code/radio-app/Cargo.toml
+++ b/nrf52-code/radio-app/Cargo.toml
@@ -33,3 +33,7 @@ incremental = false
 lto = "fat"
 opt-level = 3
 overflow-checks = false
+
+[patch.crates-io]
+defmt.git = "https://github.com/knurling-rs/defmt"
+defmt.branch = "release"
```

followed by

`cargo update -p defmt`

Everything still seemed to build OK.

**Comment by Urhengulas** [comment]
Blocked on #883 and #884.

**Comment by Urhengulas** [comment]
This is ready for review and release. As soon as you approve, @jonathanpallant, I will release the crates. Then we can update the probe-rs PR to use the crates-io version.

File: CHANGELOG.md
Comment (jonathanpallant) [inline CHANGELOG.md:63]: Why did this change a level? It feels like it was a release of some crates in and of itself, even if the main defmt crate wasn't bumped.

Comment (Urhengulas) [reply CHANGELOG.md:63]: I thought it makes sense to highlight the defmt releases more than the related releases. But we can change it back if you don't like it.

Comment (jonathanpallant) [reply CHANGELOG.md:63]: to me, it implied that the `###` releases belonged to their parent `##` release, which didn't seem right

Comment (jonathanpallant) [reply CHANGELOG.md:63]: we could maybe use emojis to mark the defmt releases, which I agree are a bigger deal. No idea which one though.

Comment (Urhengulas) [reply CHANGELOG.md]: I changed it back. The defmt releases are highlighted by the link in the heading.

**Comment by jonathanpallant** [comment]
I *think* this is OK, but I want to go and manually run `cargo semver-checks` on each thing we're releasing so I fully understand what we're changing here.

**Comment by Urhengulas** [comment]
# cargo semver-checks report

## `main` branch

```console
$ cargo semver-checks -p defmt-parser -p defmt-decoder --features unstable
     Parsing defmt-decoder v0.3.11 (current)
      Parsed [   9.142s] (current)
     Parsing defmt-decoder v0.3.11 (baseline)
      Parsed [   8.841s] (baseline)
    Checking defmt-decoder v0.3.11 -> v0.3.11 (no change)
     Checked [   0.015s] 94 checks: 93 pass, 1 fail, 0 warn, 0 skip

--- failure enum_marked_non_exhaustive: enum marked #[non_exhaustive] ---

Description:
A public enum has been marked #[non_exhaustive]. Pattern-matching on it outside of its crate must now include a wildcard pattern like `_`, or it will fail to compile.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#attr-adding-non-exhaustive
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/enum_marked_non_exhaustive.ron

Failed in:
  enum FormatterFormat in /home/urhengulas/Documents/github.com/knurling-rs/defmt/decoder/src/log/format/mod.rs:276

     Summary semver requires new major version: 1 major and 0 minor checks failed
    Finished [  18.016s] defmt-decoder
     Parsing defmt-parser v0.3.4 (current)
      Parsed [   0.816s] (current)
     Parsing defmt-parser v0.3.4 (baseline, cached)
      Parsed [   0.184s] (baseline)
    Checking defmt-parser v0.3.4 -> v0.3.4 (no change)
     Checked [   0.010s] 94 checks: 93 pass, 1 fail, 0 warn, 0 skip

--- failure enum_variant_added: enum variant added on exhaustive enum ---

Description:
A publicly-visible enum without #[non_exhaustive] has a new variant.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#enum-variant-new
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/enum_variant_added.ron

Failed in:
  variant DisplayHint:Octal in /home/urhengulas/Documents/github.com/knurling-rs/defmt/parser/src/display_hint.rs:16

     Summary semver requires new major version: 1 major and 0 minor checks failed
    Finished [   1.022s] defmt-parser
$ cargo semver-checks -p defmt-print -p defmt-macros
error: no crates with library targets selected, nothing to semver-check
note: only library targets contain an API surface that can be checked for semver
note: skipped the following crates since they have no library target: defmt-macros, defmt-print
$ cargo semver-checks -p defmt --default-features
     Parsing defmt v0.3.8 (current)
      Parsed [   1.143s] (current)
     Parsing defmt v0.3.8 (baseline, cached)
      Parsed [   0.106s] (baseline)
    Checking defmt v0.3.8 -> v0.3.8 (no change)
     Checked [   0.009s] 94 checks: 94 pass, 0 skip
     Summary no semver update required
    Finished [   1.268s] defmt

```

## `release` branch

```console
$ cargo semver-checks -p defmt-parser -p defmt-decoder --features unstable
     Parsing defmt-decoder v0.4.0 (current)
      Parsed [   9.571s] (current)
     Parsing defmt-decoder v0.3.11 (baseline, cached)
      Parsed [   0.185s] (baseline)
    Checking defmt-decoder v0.3.11 -> v0.4.0 (major change)
     Checked [   0.000s] 0 checks: 0 pass, 94 skip
     Summary no semver update required
    Finished [   9.774s] defmt-decoder
     Parsing defmt-parser v0.4.0 (current)
      Parsed [   0.900s] (current)
     Parsing defmt-parser v0.3.4 (baseline, cached)
      Parsed [   0.250s] (baseline)
    Checking defmt-parser v0.3.4 -> v0.4.0 (major change)
     Checked [   0.000s] 0 checks: 0 pass, 94 skip
     Summary no semver update required
    Finished [   1.163s] defmt-parser
```

**Comment by jonathanpallant** [comment]
That looks good for "defmt-parser" and "defmt-decoder" on the host side. What about "defmt-print" on the host side and "defmt" on the target side? I think that's all four things you're proposing to publish here?

**Comment by Urhengulas** [comment]
>  What about "defmt-print" on the host side and "defmt" on the target side? I think that's all four things you're proposing to publish here?

There are no conflicts on the `main` branch already. So there are no conflicts in the `release` branch as well. I didn't include the output for brevity, as it is the same, but just added it for completeness.

**Comment by Urhengulas** [comment]
@jonathanpallant ping

**Comment by jonathanpallant** [comment]
OK looks good

**Review by jonathanpallant — APPROVED** [review]

**Comment by jonathanpallant** [comment]
@Urhengulas - does the change in ef2cef8659a9101cc2be2572b949033dce932724 mean we have to bump the Defmt Version from v4 to v5? It's a backwards compatible change but ... how should a decoder know whether this it has "V4 organised into log levels" or "V4 with logs in random order". I guess it could look for the `__DEFMT_MARKER_TRACE_START` symbol?

**Comment by jonathanpallant** [comment]
> @Urhengulas - does the change in [ef2cef8](https://github.com/knurling-rs/defmt/commit/ef2cef8659a9101cc2be2572b949033dce932724) mean we have to bump the Defmt Version from v4 to v5? It's a backwards compatible change but ... how should a decoder know whether this it has "V4 organised into log levels" or "V4 with logs in random order". I guess it could look for the `__DEFMT_MARKER_TRACE_START` symbol?

Because the binary format sent over the transport hasn't changed, and it's backwards compatible, I'm happy to keep calling this V4. It's just V4 with extra stuff, and it's trivial to see if the extra stuff is supported.

**Pull request merged**
